openlibrary/plugins/importapi/import_validator.py

```diff
@@ -21,29 +22,34 @@ class Book(BaseModel):
     publishers: list[str]
     publish_date: str
 
-    @validator("source_records", "authors", "publishers")
+    @field_validator("source_records", "authors", "publishers")
+    @classmethod
     def list_must_not_be_empty(cls, v):
-        if not v:
-            raise ValueError("Lists must not be empty")
-        return v
+        if v:
+            return v
+        raise ValueError("Lists must not be empty")
 
+    # TODO[pydantic]: We couldn't refactor the `validator`, please replace it by `field_validator` manually.
+    # Check https://docs.pydantic.dev/dev-v2/migration/#changes-to-validators for more information.
     @validator("source_records", "publishers", each_item=True)
```

Comment:
@validator and each_item have both been removed in Pydantic v2.
https://docs.pydantic.dev/dev-v2/migration/#changes-to-validators

Reply:
I guess the cutest way to do what you want right now, would be:

```py
from typing import TypeVar, Annotated

from annotated_types import MinLen
from pydantic import BaseModel, ValidationError

T = TypeVar("T")

NonEmptyList = Annotated[list[T], MinLen(1)]
NonEmptyStr = Annotated[str, MinLen(1)]


class Book(BaseModel):
    title: str
    source_records: NonEmptyList[NonEmptyStr]


print(Book(title="foo", source_records=["bar"]))

try:
    Book(title="foo", source_records=[])
except ValidationError as e:
    print(e.errors())

try:
    Book(title="foo", source_records=[""])
except ValidationError as e:
    print(e.errors())
```

Reply:
Very slick!  Thanks massively!